### PR TITLE
Tweak the Printer code in runtime for smaller code

### PR DESCRIPTION
**Pull request by steven-johnson**
TL;DR: template expansion meant that we had more replicated code than expected from the inline expansion of code in Printer and friends. Restructured to try to make the call sites as small as possible. It's a modest code-size savings but nonzero.

(In hindsight, I could have done most of this by just adding NEVER_INLINE to select parts of the existing code, but the refactoring into a non-template base felt like cleaner code; we still need the NEVER_INLINE in various places because the Clang inliner is smarter than us, and is preferring speed to code size with aggressive inlining, but 99% of the time we want to use Printer in the runtime is not speed critical.)

### CODE REVIEW

**Comment by steven-johnson** [comment]
Bah, too much trouble to get linking to work correctly. Closing.
